pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+	mcpclient "istio.io/istio/pkg/mcp/client"
+)
+
+var errUnsupported = errors.New("this operation is not supported by mcp controller")
+
+// CoreDataModel is a combined interface for ConfigStoreCache
+// MCP Updater and ServiceDiscovery
+type CoreDataModel interface {
+	model.ConfigStoreCache
+	mcpclient.Updater
+}
+
+// Controller is a temporary storage for the changes received
+// via MCP server
+type Controller struct {
+	configStoreMu sync.RWMutex
+	// keys [type][namespace][name]
+	configStore              map[string]map[string]map[string]model.Config
+	descriptorsByMessageName map[string]model.ProtoSchema
+}
+
+// NewController provides a new CoreDataModel controller
+func NewController() CoreDataModel {
+	descriptorsByMessageName := make(map[string]model.ProtoSchema, len(model.IstioConfigTypes))
+	for _, descriptor := range model.IstioConfigTypes {
+		descriptorsByMessageName[descriptor.MessageName] = descriptor
+	}
+
+	// Remove this when https://github.com/istio/istio/issues/7947 is done
+	configStore := make(map[string]map[string]map[string]model.Config)
+	for _, typ := range model.IstioConfigTypes.Types() {
+		configStore[typ] = make(map[string]map[string]model.Config)
+	}
+	return &Controller{
+		configStore:              configStore,
+		descriptorsByMessageName: descriptorsByMessageName,
+	}
+}
+
+// ConfigDescriptor returns all the ConfigDescriptors that this
+// controller is responsible for
+func (c *Controller) ConfigDescriptor() model.ConfigDescriptor {
+	return model.IstioConfigTypes
+}
+
+// List returns all the config that is stored by type and namespace
+// if namespace is empty string it returns config for all the namespaces
+func (c *Controller) List(typ, namespace string) (out []model.Config, err error) {
+	_, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return nil, fmt.Errorf("list unknown type %s", typ)
+	}
+	c.configStoreMu.Lock()
+	data, exists := c.configStore[typ]
+	c.configStoreMu.Unlock()
+	if exists && len(data) == 0 {
+		return out, nil
+	}
```

Comment:
This extra check isn't really needed. The range clauses below handled things properly. Also, you can collapse the locking into a single access and reuse the `byType` submap when iterating the namespaces. 

```golang
c.configStoreMu.Lock()
byType, ok := c.configStore[typ]
c.configStoreMu.Unlock()
if !ok {
    return nil, nil
}

// byType does not need locking since we 
// automatically replace the entire sub-map

var out []model.Config

if namespace == "" {
    for _, byNamespace := range byType {
	    for _, config := range byNamespace {
			out = append(out, config)
		}
	}
	return out, nil
}

if byNamespace, ok := data[namespace]; ok {
	for _, config := range byNamespace {
		out = append(out, config)
	}
}

return out, nil
```

Reply:
It makes sense to lock the configStore once and re-use the byType, cleaned up.